pdns/dnsdistdist/dnsdist-tcp-downstream.cc

```diff
@@ -0,0 +1,543 @@
+  }
+  else {
+    // store query in the list of queries to send
+    d_pendingQueries.push_back(std::move(query));
+  }
+}
+
+bool TCPConnectionToBackend::reconnect()
+{
+  std::unique_ptr<Socket> result;
+
+  if (d_socket) {
+    DEBUGLOG("closing socket "<<d_socket->getHandle());
+    shutdown(d_socket->getHandle(), SHUT_RDWR);
+    d_socket.reset();
+    d_ioState.reset();
+    --d_ds->tcpCurrentConnections;
+  }
+
+  d_fresh = true;
+
+  do {
+    vinfolog("TCP connecting to downstream %s (%d)", d_ds->getNameWithAddr(), d_downstreamFailures);
+    DEBUGLOG("Opening TCP connection to backend "<<d_ds->getNameWithAddr());
+    try {
+      result = std::unique_ptr<Socket>(new Socket(d_ds->remote.sin4.sin_family, SOCK_STREAM, 0));
+      DEBUGLOG("result of connect is "<<result->getHandle());
+
+      if (!IsAnyAddress(d_ds->sourceAddr)) {
+        SSetsockopt(result->getHandle(), SOL_SOCKET, SO_REUSEADDR, 1);
+#ifdef IP_BIND_ADDRESS_NO_PORT
+        if (d_ds->ipBindAddrNoPort) {
+          SSetsockopt(result->getHandle(), SOL_IP, IP_BIND_ADDRESS_NO_PORT, 1);
+        }
+#endif
+#ifdef SO_BINDTODEVICE
+        if (!d_ds->sourceItfName.empty()) {
+          int res = setsockopt(result->getHandle(), SOL_SOCKET, SO_BINDTODEVICE, d_ds->sourceItfName.c_str(), d_ds->sourceItfName.length());
+          if (res != 0) {
+            vinfolog("Error setting up the interface on backend TCP socket '%s': %s", d_ds->getNameWithAddr(), stringerror());
+          }
+        }
+#endif
+        result->bind(d_ds->sourceAddr, false);
+      }
+      result->setNonBlocking();
+#ifdef MSG_FASTOPEN
+      if (!d_ds->tcpFastOpen || !isFastOpenEnabled()) {
+        SConnectWithTimeout(result->getHandle(), d_ds->remote, /* no timeout, we will handle it ourselves */ 0);
+      }
+#else
+      SConnectWithTimeout(result->getHandle(), d_ds->remote, /* no timeout, we will handle it ourselves */ 0);
+#endif /* MSG_FASTOPEN */
+
+      d_socket = std::move(result);
+      ++d_ds->tcpCurrentConnections;
+      return true;
+    }
+    catch(const std::runtime_error& e) {
+      vinfolog("Connection to downstream server %s failed: %s", d_ds->getName(), e.what());
+      d_downstreamFailures++;
+      if (d_downstreamFailures > d_ds->retries) {
+        throw;
+      }
+    }
+  }
+  while (d_downstreamFailures <= d_ds->retries);
+
+  return false;
+}
+
+void TCPConnectionToBackend::handleTimeout(const struct timeval& now, bool write)
+{
+  /* in some cases we could retry, here, reconnecting and sending our pending responses again */
+  if (write) {
+    ++d_ds->tcpWriteTimeouts;
+  }
+  else {
+    ++d_ds->tcpReadTimeouts;
+  }
+
+  if (d_ioState) {
+    d_ioState->reset();
+  }
+
+  notifyAllQueriesFailed(now, FailureReason::timeout);
+}
+
+void TCPConnectionToBackend::notifyAllQueriesFailed(const struct timeval& now, FailureReason reason)
+{
+  d_connectionDied = true;
+
+  if (!d_usedForXFR) {
+    d_ds->outstanding -= d_pendingResponses.size();
```

Comment:
Maybe a >= d_pendingResponses.size() check here to detect accounting errors wrt outstanding?